Data validation

In [2]:
#import pandas for data manipulation and analysis
import pandas as pd
#import numpy for numerical operations
import numpy as np

In [4]:
#load the raw retail sales dataset into a pandas DataFrame
df=pd.read_csv(r"C:\Users\dell\Documents\train.csv")
#display the first five rows of the dataset
df.head()

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,State,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales
0,1,CA-2017-152156,08/11/2017,11/11/2017,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,Kentucky,42420.0,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600
1,2,CA-2017-152156,08/11/2017,11/11/2017,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,Kentucky,42420.0,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400
2,3,CA-2017-138688,12/06/2017,16/06/2017,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,California,90036.0,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200
3,4,US-2016-108966,11/10/2016,18/10/2016,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,Florida,33311.0,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775
4,5,US-2016-108966,11/10/2016,18/10/2016,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,Florida,33311.0,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680


In [5]:
#check the number of rows and columns in the dataset
df.shape

(9800, 18)

In [7]:
#display all column names in the dataset
df.columns

Index(['Row ID', 'Order ID', 'Order Date', 'Ship Date', 'Ship Mode',
       'Customer ID', 'Customer Name', 'Segment', 'Country', 'City', 'State',
       'Postal Code', 'Region', 'Product ID', 'Category', 'Sub-Category',
       'Product Name', 'Sales'],
      dtype='object')

In [6]:
#display information about the data, including data types and the number of non-null values in each column
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9800 entries, 0 to 9799
Data columns (total 18 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Row ID         9800 non-null   int64  
 1   Order ID       9800 non-null   object 
 2   Order Date     9800 non-null   object 
 3   Ship Date      9800 non-null   object 
 4   Ship Mode      9800 non-null   object 
 5   Customer ID    9800 non-null   object 
 6   Customer Name  9800 non-null   object 
 7   Segment        9800 non-null   object 
 8   Country        9800 non-null   object 
 9   City           9800 non-null   object 
 10  State          9800 non-null   object 
 11  Postal Code    9789 non-null   float64
 12  Region         9800 non-null   object 
 13  Product ID     9800 non-null   object 
 14  Category       9800 non-null   object 
 15  Sub-Category   9800 non-null   object 
 16  Product Name   9800 non-null   object 
 17  Sales          9800 non-null   float64
dtypes: float

In [11]:
#generate discriptiive statistics for numerical columns
df.describe()

,Row ID,Postal Code,Sales
count,9800.000000,9789.000000,9800.000000
mean,4900.500000,55273.322403,230.769059
std,2829.160653,32041.223413,626.651875
min,1.000000,1040.000000,0.444000
25%,2450.750000,23223.000000,17.248000
50%,4900.500000,58103.000000,54.490000
75%,7350.250000,90008.000000,210.605000
max,9800.000000,99301.000000,22638.480000


In [8]:
#count the number of missing values in each column
df.isnull().sum()

Row ID            0
Order ID          0
Order Date        0
Ship Date         0
Ship Mode         0
Customer ID       0
Customer Name     0
Segment           0
Country           0
City              0
State             0
Postal Code      11
Region            0
Product ID        0
Category          0
Sub-Category      0
Product Name      0
Sales             0
dtype: int64

In [10]:
df.duplicated().sum()

0

In [12]:
df.dtypes

Row ID             int64
Order ID          object
Order Date        object
Ship Date         object
Ship Mode         object
Customer ID       object
Customer Name     object
Segment           object
Country           object
City              object
State             object
Postal Code      float64
Region            object
Product ID        object
Category          object
Sub-Category      object
Product Name      object
Sales            float64
dtype: object

In [14]:
#display the unique values in the ship mode column
df["Ship Mode"].unique()

array(['Second Class', 'Standard Class', 'First Class', 'Same Day'],
      dtype=object)

In [16]:
#display the unique values in the segment column
df["Segment"].unique()

array(['Consumer', 'Corporate', 'Home Office'], dtype=object)

In [18]:
#display the unique values in the Region column
df["Region"].unique()

array(['South', 'West', 'Central', 'East'], dtype=object)

In [20]:
#display the unique values in the category column
df["Category"].unique()

array(['Furniture', 'Office Supplies', 'Technology'], dtype=object)

In [22]:
#display the unique values in the sub-category column
df["Sub-Category"].unique()

array(['Bookcases', 'Chairs', 'Labels', 'Tables', 'Storage',
       'Furnishings', 'Art', 'Phones', 'Binders', 'Appliances', 'Paper',
       'Accessories', 'Envelopes', 'Fasteners', 'Supplies', 'Machines',
       'Copiers'], dtype=object)

In [24]:
#count the number of records in each ship mode
df["Ship Mode"].value_counts()

Ship Mode
Standard Class    5859
Second Class      1902
First Class       1501
Same Day           538
Name: count, dtype: int64

In [26]:
#count the number of records in each customer segment
df["Segment"].value_counts()

Segment
Consumer       5101
Corporate      2953
Home Office    1746
Name: count, dtype: int64

In [28]:
#count the number of records in each region
df["Region"].value_counts()

Region
West       3140
East       2785
Central    2277
South      1598
Name: count, dtype: int64

In [30]:
#count the number of records in each product category
df["Category"].value_counts()

Category
Office Supplies    5909
Furniture          2078
Technology         1813
Name: count, dtype: int64

In [32]:
#count the number of records in each sub-Category
df["Sub-Category"].value_counts()

Sub-Category
Binders        1492
Paper          1338
Furnishings     931
Phones          876
Storage         832
Art             785
Accessories     756
Chairs          607
Appliances      459
Labels          357
Tables          314
Envelopes       248
Bookcases       226
Fasteners       214
Supplies        184
Machines        115
Copiers          66
Name: count, dtype: int64

In [34]:
#define the main identifier columns
id_columns = ["Row ID", "Order ID", "Customer ID", "Product ID"]
#check the number of unique values in each identifier column
for column in id_columns:
    print(f"{column}: {df[column].nunique()} unique values")

Row ID: 9800 unique values
Order ID: 4922 unique values
Customer ID: 793 unique values
Product ID: 1861 unique values


In [36]:
#identify records with negative sales values
negative_sales = df[df["Sales"] < 0 ]
#display the number of negative sales records
len(negative_sales)

0

In [38]:
#convert order data and ship date to datetime format. format="mixed" allows pandas to correctly handle different date formats. dayfirst=True  tells pandas that the first number represents the day
df["Order Date"] = pd.to_datetime(df["Order Date"], format = "mixed", dayfirst=True)
df["Ship Date"] = pd.to_datetime(df["Ship Date"], format = "mixed", dayfirst=True)
#check that both data columns have been converted to datetime
df[["Order Date", "Ship Date"]].dtypes

Order Date    datetime64[ns]
Ship Date     datetime64[ns]
dtype: object

In [57]:
#display the records where postal code is missing
missing_postel = df[df["Postal Code"].isnull()]
#display relevant information about the missing postal code records
missing_postel[["Order ID","City","State","Postal Code","Region"]]

,Order ID,City,State,Postal Code,Region
2234,CA-2018-104066,Burlington,Vermont,NaN,East
5274,CA-2016-162887,Burlington,Vermont,NaN,East
8798,US-2017-150140,Burlington,Vermont,NaN,East
9146,US-2017-165505,Burlington,Vermont,NaN,East
9147,US-2017-165505,Burlington,Vermont,NaN,East
9148,US-2017-165505,Burlington,Vermont,NaN,East
9386,US-2018-127292,Burlington,Vermont,NaN,East
9387,US-2018-127292,Burlington,Vermont,NaN,East
9388,US-2018-127292,Burlington,Vermont,NaN,East
9389,US-2018-127292,Burlington,Vermont,NaN,East


In [40]:
#calculate the first quartile and third quartile of sales
q1 = df["Sales"].quantile(0.25)
q3 = df["Sales"].quantile(0.75)
#calculate the interquartile range
iqr = q3-q1
#calculate the lower and upper boundaries for potential outliers
lower_bound = q1 - 1.5 * iqr
upper_bound = q3 + 1.5 * iqr
#display the boundaries
print("Lower Bound: ", lower_bound)
print("Upper Bound: ", upper_bound)

Lower Bound:  -272.7875000000001
Upper Bound:  500.6405000000001


In [42]:
#identify potential sales outliers using the iqr method
sales_outliers = df[~df["Sales"].between(lower_bound, upper_bound)]
#count the number of potential sales outliers
sales_outlier_count = sales_outliers.shape[0]
#display the number of potential sales outliers
print("Number of potential sales outliers: ", sales_outlier_count)

Number of potential sales outliers:  1145


DATA CLEANING

In [45]:
#create a copy of the dataset for cleaning. This keeps the original DataFrame seperate from the cleaned version 
cleaned_df = df.copy()

In [47]:
#ensure order date and ship date are stored as datetime values
cleaned_df["Order Date"] = pd.to_datetime(cleaned_df["Order Date"], dayfirst=True)
cleaned_df["Ship Date"] = pd.to_datetime(cleaned_df["Ship Date"], dayfirst=True)
cleaned_df[["Order Date", "Ship Date"]].dtypes

Order Date    datetime64[ns]
Ship Date     datetime64[ns]
dtype: object

In [49]:
#identity text columns in the cleaned dataset
text_columns = cleaned_df.select_dtypes(include="object").columns
#remove leading and trailing spaces from text values
for column in text_columns:
    cleaned_df[column] = cleaned_df[column].str.strip()
cleaned_df

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,State,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales
0,1,CA-2017-152156,2017-11-08,2017-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,Kentucky,42420.0,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600
1,2,CA-2017-152156,2017-11-08,2017-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,Kentucky,42420.0,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400
2,3,CA-2017-138688,2017-06-12,2017-06-16,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,California,90036.0,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200
3,4,US-2016-108966,2016-10-11,2016-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,Florida,33311.0,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775
4,5,US-2016-108966,2016-10-11,2016-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,Florida,33311.0,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9795,9796,CA-2017-125920,2017-05-21,2017-05-28,Standard Class,SH-19975,Sally Hughsby,Corporate,United States,Chicago,Illinois,60610.0,Central,OFF-BI-10003429,Office Supplies,Binders,"Cardinal HOLDit! Binder Insert Strips,Extra St...",3.7980
9796,9797,CA-2016-128608,2016-01-12,2016-01-17,Standard Class,CS-12490,Cindy Schnelling,Corporate,United States,Toledo,Ohio,43615.0,East,OFF-AR-10001374,Office Supplies,Art,"BIC Brite Liner Highlighters, Chisel Tip",10.3680
9797,9798,CA-2016-128608,2016-01-12,2016-01-17,Standard Class,CS-12490,Cindy Schnelling,Corporate,United States,Toledo,Ohio,43615.0,East,TEC-PH-10004977,Technology,Phones,GE 30524EE4,235.1880
9798,9799,CA-2016-128608,2016-01-12,2016-01-17,Standard Class,CS-12490,Cindy Schnelling,Corporate,United States,Toledo,Ohio,43615.0,East,TEC-PH-10000912,Technology,Phones,Anker 24W Portable Micro USB Car Charger,26.3760


In [51]:
#convert postal code to a text column so it can contain both postal code the label "unknown"
cleaned_df["Postal Code"] = cleaned_df["Postal Code"].astype("Int64").astype("string")
#replace missing code values with "unknown"
cleaned_df["Postal Code"] = cleaned_df["Postal Code"].fillna("unknown")
#check the number of records assigned the "unknown" postal code label 
(cleaned_df["Postal Code"] == "unknown").sum()

11

Sales outlier treatment: The IQR method identified 1,145 potential sales outliers. These observations were retained because statistical outliers are not necessarily data errors, and there was no evidence that these values were incorrect.

In [54]:
#check the final number of rows and columns
print("Dataset shape: ", cleaned_df.shape)
#check for duplicate rows
print("Duplicate rows: ", cleaned_df.duplicated().sum())
#check for missing values
print("Missing values: ", cleaned_df.isnull().sum())
#check the date column date types
print("Date data types: ", cleaned_df[["Order Date","Ship Date"]].dtypes)

Dataset shape:  (9800, 18)
Duplicate rows:  0
Missing values:  Row ID           0
Order ID         0
Order Date       0
Ship Date        0
Ship Mode        0
Customer ID      0
Customer Name    0
Segment          0
Country          0
City             0
State            0
Postal Code      0
Region           0
Product ID       0
Category         0
Sub-Category     0
Product Name     0
Sales            0
dtype: int64
Date data types:  Order Date    datetime64[ns]
Ship Date     datetime64[ns]
dtype: object


In [56]:
#compare the number of rows and columns before and after cleaning 
print("Original dataset: ", df.shape)
print("Cleaned dataset: ", cleaned_df.shape)

Original dataset:  (9800, 18)
Cleaned dataset:  (9800, 18)


In [58]:
#display the first five rows of the cleaned dataset
cleaned_df.head()

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,State,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales
0,1,CA-2017-152156,2017-11-08,2017-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,Kentucky,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600
1,2,CA-2017-152156,2017-11-08,2017-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,Kentucky,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400
2,3,CA-2017-138688,2017-06-12,2017-06-16,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,California,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200
3,4,US-2016-108966,2016-10-11,2016-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,Florida,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775
4,5,US-2016-108966,2016-10-11,2016-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,Florida,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680


In [60]:
#Save the cleaned dataset as a new csv file. The original train.csv file remains unchanged
cleaned_df.to_csv("cleaned_retail_sales.csv", index=False)

Data Cleaning Summary

The dataset initially contained 9,800 records and 18 columns. Data quality checks were performed to examine missing values, duplicate records, incorrect data types, categorical values, negative sales values, missing postal codes, and potential numerical outliers. 

The dataset contained 11 missing postal code values. These records were retained because the other information contained in the transactions remained useful. Since the correct postal codes could not be reliably determined from the dataset, the missing values were replaced with "unknown". 

The order date and ship date columns were converted to appropriate datatime formats. Text columns were stripped of unnecessary leading and trailing spaces.

No completely duplicate rows were found and no negative value were identified. Potential sales outliers were detected using the IQR method but retained because there was no evidence that they represented data-entry errors.

After cleaning, the dataset remained at 9,800 rows and 18 columns and was saved as "cleaned_retail_sales.csv".